### PR TITLE
Update for Prospress/action-scheduler#113

### PR DESCRIPTION
**Pull request by thenbrent**
This adds a bunch of changes to make the custom tables compatible with https://github.com/Prospress/action-scheduler/pull/113. See commit messages for details.

I need to do a final review on these now that https://github.com/Prospress/action-scheduler/pull/113 is merged - most of the commits were made a while ago, but I wanted to get a PR up now so the changes are visible (and to avoid something like https://github.com/Prospress/action-scheduler-custom-tables/pull/17 happening again).

This PR still includes https://github.com/Prospress/action-scheduler-custom-tables/commit/80ade0a23873ce41c3c145c667b2e8a3a8fb4561, which is equivalent, but not as good as, https://github.com/Prospress/action-scheduler-custom-tables/pull/17 (so the patches in that PR should be used instead). I'll either revert or remove SHA: 80ade0a before proposing this for final review.

### CODE REVIEW

File: src/Migration/Migration_Scheduler.php [hunk 1]
```diff
@@ -45,6 +45,13 @@ public function mark_complete() {
 		do_action( 'action_scheduler/custom_tables/migration_complete' );
 	}
 
+	/**
+	 * @return bool Whether to do background migrations, defaults to true until the migration is complete
+	 */
+	public function do_background_migration() {
+		return apply_filters( 'action_scheduler/custom_tables/do_background_migration', ! $this->is_migration_complete() );
```
Comment (JPry) [inline src/Migration/Migration_Scheduler.php]: I think we should type-cast this to `bool` so that we can use identical comparison in the main `Plugin.php` file. (see comments in that file)

File: src/Plugin.php [hunk 2]
```diff
@@ -63,20 +63,26 @@ public function register_cli_command() {
 	 * @return void
 	 */
 	public function schedule_migration() {
-		if ( ! apply_filters( 'action_scheduler/custom_tables/do_background_migration', true ) ) {
-			return;
-		}
 
 		$scheduler = new Migration\Migration_Scheduler();
-		if ( $scheduler->is_migration_complete() ) {
+		if ( false == $scheduler->do_background_migration() || $scheduler->is_migration_scheduled() ) {
```
Comment (JPry) [inline src/Plugin.php]: This should be `===` for comparison.

File: src/Plugin.php [hunk 3]
```diff
@@ -63,20 +63,26 @@ public function register_cli_command() {
...
-		if ( $scheduler->is_migration_scheduled() ) {
+	/**
+	 * Attach the callback to run the background migration process
+	 *
+	 * @return void
+	 */
+	public function hook_scheduled_migration() {
+
+		$scheduler = new Migration\Migration_Scheduler();
+		if ( false == $scheduler->do_background_migration() ) {
```
Comment (JPry) [inline src/Plugin.php]: This should also be `===` for comparison

File: src/Plugin.php [hunk 4]
```diff
@@ -103,17 +109,20 @@ public function hook_admin_notices() {
...
+
 		add_filter( 'action_scheduler_store_class', [ $this, 'set_store_class' ], 10, 1 );
 		add_filter( 'action_scheduler_logger_class', [ $this, 'set_logger_class' ], 10, 1 );
 		add_action( 'plugins_loaded', [ $this, 'register_cli_command' ], 10, 0 );
+		add_action( 'plugins_loaded', [ $this, 'hook_scheduled_migration' ], 1000, 0 );
 		add_action( 'shutdown', [ $this, 'schedule_migration' ], 0, 0 );
 
-		// TODO: add this to the correct load-* hook once the admin page exists
-		add_action( 'load-action-scheduler', [ $this, 'hook_admin_notices' ], 10, 0 );
+		// Action Scheduler may be displayed as a Tools screen or WooCommerce > Status adminstration screen
```
Comment (JPry) [inline src/Plugin.php]: `s/adminstration/administration`

**Comment by thenbrent** [comment]
Thanks for the review @JPry. I've fixed up all of those issues!

**Comment by JPry** [comment]
@thenbrent I've made some updates to the tests, but I'm still seeing some failures I can't explain. I have the `master` branch of Action Scheduler running alongside this particular branch. Would you mind taking a look at the failing tests, as well as the changes I made? Here's the output I'm seeing:

```
phpunit@5.7 -c tests/phpunit.xml.dist                  
Installing...
Running as single site... To run multisite, use -c tests/phpunit/multisite.xml
Not running ajax tests. To execute these, use --group ajax.
Not running ms-files tests. To execute these, use --group ms-files.
Not running external-http tests. To execute these, use --group external-http.
PHPUnit 5.7.27 by Sebastian Bergmann and contributors.

...............................................................  63 / 188 ( 33%)
.....................FF........................FF.............. 126 / 188 ( 67%)
..............................................................  188 / 188 (100%)

Time: 3.83 seconds, Memory: 14.00MB

There were 4 failures:

1) Action_Scheduler\Custom_Tables\Migration_Scheduler_Test::test_scheduler_runs_migration
Failed asserting that actual size 10 matches expected size 15.

/Users/jpry/Sites/wordpress-develop/src/wp-content/plugins/action-scheduler-custom-tables/tests/phpunit/migration/Migration_Scheduler_Test.php:74
/Users/jpry/Sites/wordpress-develop/src/wp-content/plugins/action-scheduler-custom-tables/tests/UnitTestCase.php:43

2) Action_Scheduler\Custom_Tables\Migration_Scheduler_Test::test_scheduler_runs_migration
Failed asserting that actual size 10 matches expected size 15.

/Users/jpry/Sites/wordpress-develop/src/wp-content/plugins/action-scheduler-custom-tables/tests/phpunit/migration/Migration_Scheduler_Test.php:74
/Users/jpry/Sites/wordpress-develop/src/wp-content/plugins/action-scheduler-custom-tables/tests/UnitTestCase.php:46

3) procedural_api_Test::test_unschedule
Failed asserting that two strings are equal.
--- Expected
+++ Actual
@@ @@
-'59b0c42b95d8111e1786c61364fdc305'
+''

/Users/jpry/Sites/wordpress-develop/src/wp-content/plugins/action-scheduler/tests/phpunit/procedural_api/procedural_api_Test.php:67
/Users/jpry/Sites/wordpress-develop/src/wp-content/plugins/action-scheduler/tests/phpunit/deprecated/ActionScheduler_UnitTestCase.php:35

4) procedural_api_Test::test_unschedule
Failed asserting that two strings are equal.
--- Expected
+++ Actual
@@ @@
-'7ab22e9d773cddfc1b5f11ec2027f5e0'
+''

/Users/jpry/Sites/wordpress-develop/src/wp-content/plugins/action-scheduler/tests/phpunit/procedural_api/procedural_api_Test.php:67
/Users/jpry/Sites/wordpress-develop/src/wp-content/plugins/action-scheduler/tests/phpunit/deprecated/ActionScheduler_UnitTestCase.php:38

FAILURES!
Tests: 564, Assertions: 518, Failures: 4.
```

**Comment by JPry** [comment]
@thenbrent I finally [found the issue](https://github.com/Prospress/action-scheduler-custom-tables/pull/19/commits/15710ce3798369b54502181aaa0459919990e475) that was causing the tests to fail, and everything is passing now. I'm marking this approved, and I'll leave it to you to merge since I added more, starting with https://github.com/Prospress/action-scheduler-custom-tables/pull/19/commits/bf4ea66495b6e558dd3748167cdc44655c3c23e5.

**Comment by JPry** [comment]
I went back to work on #17 after this, and it looks like any new work is going to depend on this PR. So as not to hold anything up, I'm going to go ahead and merge this. @thenbrent feel free to review the commits I added, and if necessary we can create a new PR to change anything. I'll leave the branch available for now instead of deleting it.

**Comment by thenbrent** [comment]
> feel free to review the commits I added

Those all LGTM, thanks @JPry 🙇
